### ISSUE TITLE
Calling a function that returns never isn't treated as stopping control flow

### ISSUE THREAD

**Issue by marijnh** [microsoft/TypeScript#32677]
**Code**

```ts
function raise(msg: string): never { throw new Error(msg) }

let x = 1 as (string | number)

if (typeof x != "number") raise("Stop")

console.log(x * 2)
```

**Expected behavior:**

I would expect the control flow analysis to treat the `console.log` as only being reachable if the `if` branch wasn't taken, which means that `x` is a number, so the code is accepted. This does happen when you inline the `throw`.

**Actual behavior:**

The type of `x` isn't narrowed and an error is reported for `x * 2`.

**Playground Link:** [here](https://www.typescriptlang.org/play/#code/GYVwdgxgLglg9mABAJwIYwM4FMAUBbDAcwC5EMpkYxCBKUsLANy2UQG9EoALZOAd0QMBAUWS9k+IjUQBfAFByANliiIAHogC8iAIyJUGRDnKVqiAD6CQeAEYsaCmMCNQAngAcscZxoCE2gCIwaztkAOk0TFwAgGUoOHdwhQgEDDhlADpFOEIcDQAqRAAmGiA)

**Comment by nattthebear** [comment]
Duplicate of https://github.com/microsoft/TypeScript/issues/16546, more or less.

**Comment by marijnh** [comment]
Ah, didn't find that. But that one wasn't ever resolved and closed for housekeeping reasons. The `return` solution mostly works, but is awkward and adds extra typechecking-related cruft to the output JavaScript, so I'm using it, but am not too happy with it.

Is there a technical reason for the way this currently works?

**Comment by jack-williams** [comment]
Duplicate of #8655.

> Is there a technical reason for the way this currently works?

Read [here](https://github.com/microsoft/TypeScript/issues/8655#issuecomment-412685082).

**Comment by fatcerberus** [comment]
One thing that the above linked comment doesn’t mention is that the control flow graph is constructed *before type information is available*, so effectively this would require *every single function call* to be added to it.  As you can imagine that would be prohibitively expensive to do.

**Comment by jack-williams** [comment]
It sorta does, just not in those words.

> The control flow graph used to determine which expressions change the type of other expressions is constructed _syntactically_.

**Comment by fatcerberus** [comment]
That’s misleading though because if the return type is explicitly annotated then it’s part of the AST (and therefore still syntactical).  TS just hasn’t constructed the necessary type information internally for that to matter.

**Comment by marijnh** [comment]
I, for one, would be perfectly satisfied with this only working for functions explicitly annotated to return `never`. Would that still run into the control flow graph performance issue you mentioned?

**Comment by dragomirtitian** [comment]
@marijnh yes it would still have the same performance issues. 

At the time the control flow graph is created the compiler does not know the return type of a function, it just sees a call. So you either include all calls in the graph or no calls. 

All you can use to decide if a node is to be included is the syntax.

**Comment by acutmore** [comment]
Now that we have https://github.com/microsoft/TypeScript/pull/32695 - I wasn't expecting a different between the following two snippets:

```ts
function throws(): never {
  throw new Error();
}
throws();
console.log('here'); // ⚠unreachable code detected
```

However:

```ts
function throws(): never {
  throw new Error();
}
let n: string = throws();
console.log(n); // no issues here 
```

Is there somewhere that details why this is?

https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABFAFgJzgdwM4AoCUAXImAKYBupaiA3gFCLLpYmmaICiaGaBA3HQC+dOgBtSUEogC8iAOQoqpOQLoB6NVNmoMOfog2JwEOAFtTpMJIvrNOrHnx8DmuNRt0TYbHHEA6UTgAc1wwfDogA
